nodestream/schema/migrations/migrations.py

```diff
@@ -0,0 +1,160 @@
+        )
+
+    @classmethod
+    def describe_yaml_schema(cls):
+        from schema import Schema
+
+        return Schema(
+            {
+                "name": str,
+                "operations": [Operation.describe_yaml_schema()],
+                "dependencies": [str],
+            }
+        )
+
+    def is_root_migration(self) -> bool:
+        """Check if this migration is a root migration.
+
+        Returns:
+            True if this migration is a root migration, False otherwise.
+        """
+        return len(self.dependencies) == 0
+
+    def is_leaf_migration(self, graph: "MigrationGraph") -> bool:
+        """Check if this migration is a leaf migration.
+
+        Returns:
+            True if this migration is a leaf migration, False otherwise.
+        """
+        dependents = (m for m in graph.all_migrations() if self.name in m.dependencies)
+        return not any(dependents)
+
+    def write_to_file_with_default_name(self, directory: Path) -> Path:
+        """Save the migration to a directory with a default name.
+
+        Args:
+            directory: The directory to save the migration to.
+        """
+        path = directory / f"{self.name}.yaml"
+        self.write_to_file(path)
+        return path
+
+
+@dataclass(frozen=True, slots=True)
+class MigrationGraph:
+    """A directed acyclic graph of migrations.
+
+    The graph is represented as an adjacency list.
+    """
+
+    migrations_by_name: Dict[str, Migration]
+
+    def get_migration(self, name: str) -> Migration:
+        """Get a migration by name.
+
+        Args:
+            name: The name of the migration to get.
+
+        Returns:
+            The migration with the given name.
+
+        Raises:
+            KeyError: If there is no migration with the given name.
+        """
+        return self.migrations_by_name[name]
+
+    def get_ordered_migration_plan(self) -> List[Migration]:
+        plan_order = []
+
+        for migration in self.all_migrations():
+            for required_migration in self.plan_to_execute_migration(migration):
+                if required_migration not in plan_order:
+                    plan_order.append(required_migration)
+
+        return plan_order
+
+    def _iterative_dfs_traversal(self, start_node: Migration) -> List[Migration]:
+        visited_order = []
+        visited_set = set()
+        stack = [(start_node, False)]
+
+        while stack:
+            node, processed = stack.pop()
+            if node.name in visited_set:
+                continue
+            if processed:
+                visited_order.append(node)
+                visited_set.add(node.name)
+                continue
+
+            stack.append((node, True))
+            stack.extend(
+                ((self.migrations_by_name[n], False) for n in node.dependencies)
+            )
+
+        return visited_order
+
+    def plan_to_execute_migration(self, migration: Migration) -> List[Migration]:
```

Comment:
This method calling a function with the same contract sticks out to me

Reply:
I am fine removing it if you prefer. Just let me know. But for context, the intent was to provide an external method that was communicative. But internally I wanted to point out that it was  an `iterative dfs traversal` which was happening.

Reply:
Thats what I figured, in that case i'm fine with it :+1: